modules/packages/freebsd_ports

```diff
@@ -0,0 +1,74 @@
+#!/bin/sh -e
+
+command=$1
+
+while read -r line; do
+  export INPUT_$line
+done
+
+list_updates() {
+  pkg version -v -l "<" | sed -e 's/\([^ ]*\)-[^-]* .* \(.*\))/Name=\1\
+Version=\2\
+Architecture=none/'
+}
+
+case $command in
+  supports-api-version)
+    echo 1
+  ;;
+  get-package-data)
+    name="${INPUT_File?File must be given to get-package-data}"
+    echo PackageType=repo
+    echo Name=$name
+  ;;
+  list-installed)
+    pkg info -f -a | egrep '^(Name|Version|Architecture)' | sed -e 's/[ ]*:[ ]*/=/' -e 's/^Architecture=.*/Architecture=none/'
+  ;;
+  repo-install)
+    name="${INPUT_Name?Name must be given to repo-install}"
+    version="${INPUT_Version}"
+
+    export BATCH=1
+    PORT_PATH="`whereis -sq \"$name\"`"
+
+    if [ -z "$PORT_PATH" ]
+    then
+      echo "ErrorMessage=Could not install $name, port does not exist"
+      exit 0
+    fi
+
+    cd "$PORT_PATH"
+
+    if [ "$version" ]
+    then
+      available=`make -V PORTVERSION`
+      if [ "$available" != "$version" ]
+      then
+        echo "ErrorMessage=Could not install $name $version, available version was $available"
+        exit 0
+      fi
+    fi
+
+    if `pkg info -e "$name"`
+    then
+      make deinstall reinstall >&2
+    else
+      make install >&2
+    fi
+  ;;
+  list-updates)
+    portsnap --interactive fetch update >&2
+    list_updates
+  ;;
+  list-updates-local)
+    list_updates
+  ;;
+  remove)
+    name="${INPUT_Name?Name must be given to remove}"
+    BATCH=1
+    cd `whereis -sq "$name"`
+    make deinstall >&2
```

Comment:
There needs to be an explicit test and bailout here in failure scenarios. If make deinstall fails, it can ripple and break the whole system.